File: modules/mopro/ios/MoproModule.swift
import ExpoModulesCore

func convertCircomProof(proof: CircomProof) -> ExpoProof {
  let a = ExpoG1()
  a.x = proof.a.x
  a.y = proof.a.y
  a.z = proof.a.z

  let b = ExpoG2()
  b.x = proof.b.x
  b.y = proof.b.y
  b.z = proof.b.z

  let c = ExpoG1()
  c.x = proof.c.x
  c.y = proof.c.y
  c.z = proof.c.z

  let expoProof = ExpoProof()
  expoProof.a = a
  expoProof.b = b
  expoProof.c = c
  expoProof.protocol = proof.protocol
  expoProof.curve = proof.curve
  return expoProof
}

func convertCircomProofResult(proofResult: ExpoCircomProofResult) -> CircomProofResult {
  guard let proof = proofResult.proof,
    let a = proof.a,
    let b = proof.b,
    let c = proof.c,
    let inputs = proofResult.inputs,
    let `protocol` = proof.protocol,
    let curve = proof.curve
  else {
    fatalError("Invalid proof result")
  }

  let g1a = G1(x: a.x ?? "0", y: a.y ?? "0", z: a.z ?? "1")
  let g2b = G2(x: b.x ?? ["1", "0"], y: b.y ?? ["1", "0"], z: b.z ?? ["1", "0"])
  let g1c = G1(x: c.x ?? "0", y: c.y ?? "0", z: c.z ?? "1")

  let circomProof = CircomProof(
    a: g1a, b: g2b, c: g1c, protocol: `protocol`, curve: curve)
  let circomProofResult = CircomProofResult(proof: circomProof, inputs: inputs)
  return circomProofResult
}

enum CircomError: Error {
  case circomProofGenerationFailed(String)
  case circomProofVerificationFailed(String)
}

enum Halo2Error: Error {
  case halo2ProofGenerationFailed(String)
  case halo2ProofVerificationFailed(String)
}

enum NoirError: Error {
  case noirProofGenerationFailed(String)
  case noirProofVerificationFailed(String)
}

public class MoproModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('Mopro')` in JavaScript.
    Name("Mopro")

    // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
    Constants([
      "PI": Double.pi
    ])

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("generateCircomProof") {
      (zkeyPath: String, circuitInputs: String, expoProofLib: ExpoCircomProofLib) -> ExpoCircomProofResult in

      do {
        let proofLib = expoProofLib.proofLib == ProofLibOption.arkworks ? ProofLib.arkworks : ProofLib.rapidsnark
        let res = try generateCircomProof(
          zkeyPath: zkeyPath, circuitInputs: circuitInputs, proofLib: proofLib)
        let result = ExpoCircomProofResult()
        result.inputs = res.inputs
        result.proof = convertCircomProof(proof: res.proof)
        return result
      } catch {
        throw CircomError.circomProofGenerationFailed(error.localizedDescription)
      }
    }

    AsyncFunction("verifyCircomProof") {
      (zkeyPath: String, proofResult: ExpoCircomProofResult, proofLib: ExpoCircomProofLib) -> Bool in

      do {
        let proofLib = proofLib.proofLib == .arkworks ? ProofLib.arkworks : ProofLib.rapidsnark
        let isValid = try verifyCircomProof(
          zkeyPath: zkeyPath,
          proofResult: convertCircomProofResult(proofResult: proofResult),
          proofLib: ProofLib.arkworks
        )
        return isValid
      } catch {
        throw CircomError.circomProofVerificationFailed(error.localizedDescription)
      }
    }

    AsyncFunction("generateHalo2Proof") {
      (srsPath: String, pkPath: String, circuitInputs: [String: [String]]) throws -> ExpoHalo2ProofResult
      in

      do {
        let res = try generateHalo2Proof(
          srsPath: srsPath, pkPath: pkPath, circuitInputs: circuitInputs)
        let result = ExpoHalo2ProofResult()
        result.inputs = res.inputs
        result.proof = res.proof
        return result
      } catch {
        throw Halo2Error.halo2ProofGenerationFailed(error.localizedDescription)
      }
    }

    AsyncFunction("verifyHalo2Proof") {
      (srsPath: String, vkPath: String, proof: Data, publicInput: Data) throws -> Bool in

      do {
        let isValid = try verifyHalo2Proof(
          srsPath: srsPath, vkPath: vkPath, proof: proof, publicInput: publicInput)
        return isValid
      } catch {
        throw Halo2Error.halo2ProofVerificationFailed(error.localizedDescription)
      }
    }

    AsyncFunction("generateNoirProof") {
      (circuitPath: String, srsPath: String?, inputs: [String], onChain: Bool, vk: Data, lowMemoryMode: Bool) throws -> Data in

      do {
        let res = try generateNoirProof(circuitPath: circuitPath, srsPath: srsPath, inputs: inputs, onChain: onChain, vk: vk, lowMemoryMode: lowMemoryMode)
        return res
      } catch {
        print("error", error)
        throw NoirError.noirProofGenerationFailed(error.localizedDescription)
      }
    }

    AsyncFunction("verifyNoirProof") {
      (circuitPath: String, proof: Data, onChain: Bool, vk: Data, lowMemoryMode: Bool) throws -> Bool in

      do {
        let isValid = try verifyNoirProof(circuitPath: circuitPath, proof: proof, onChain: onChain, vk: vk, lowMemoryMode: lowMemoryMode)
        return isValid
      } catch {
        throw NoirError.noirProofVerificationFailed(error.localizedDescription)
      }
    }

    AsyncFunction("getNoirVerificationKey") {
      (circuitPath: String, srsPath: String?, onChain: Bool, lowMemoryMode: Bool) throws -> Data in

      do {
        let vk = try getNoirVerificationKey(circuitPath: circuitPath, srsPath: srsPath, onChain: onChain, lowMemoryMode: lowMemoryMode)
        return vk
      } catch {
        throw NoirError.noirProofGenerationFailed(error.localizedDescription)
      }
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("setValueAsync") { (value: String) in
      // Send an event to JavaScript.
      self.sendEvent(
        "onChange",
        [
          "value": value
        ])
    }

    // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // view definition: Prop, Events.
    View(MoproView.self) {
      // Defines a setter for the `name` prop.
      Prop("name") { (view: MoproView, prop: String) in
        print(prop)
      }
    }
  }
}

File: modules/mopro/android/src/main/java/expo/modules/mopro/MoproModule.kt
package expo.modules.mopro

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import java.io.File
import uniffi.mopro.*

fun convertCircomProof(proof: CircomProof): ExpoProof {
  var a = ExpoG1(proof.a.x, proof.a.y, proof.a.z)
  var b = ExpoG2(proof.b.x, proof.b.y, proof.b.z)
  var c = ExpoG1(proof.c.x, proof.c.y, proof.c.z)
  var output = ExpoProof(a, b, c, proof.protocol, proof.curve)
  return output
}

fun convertCircomProofResult(proofResult: ExpoCircomProofResult): CircomProofResult {
  var g1a = G1(proofResult.proof?.a?.x ?: "0", proofResult.proof?.a?.y ?: "0", proofResult.proof?.a?.z ?: "1")
  var g2b = G2(proofResult.proof?.b?.x ?: listOf("1", "0"), proofResult.proof?.b?.y ?: listOf("1", "0"), proofResult.proof?.b?.z ?: listOf("1", "0"))
  var g1c = G1(proofResult.proof?.c?.x ?: "0", proofResult.proof?.c?.y ?: "0", proofResult.proof?.c?.z ?: "1")
  var circomProof = CircomProof(g1a, g2b, g1c, proofResult.proof?.protocol ?: "groth16", proofResult.proof?.curve ?: "bn128")
  var circomProofResult = CircomProofResult(circomProof, proofResult.inputs ?: listOf("0"))
  return circomProofResult
}

class MoproModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a
    // string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for
    // clarity.
    // The module will be accessible from `requireNativeModule('Mopro')` in JavaScript.
    Name("Mopro")

    // Sets constant properties on the module. Can take a dictionary or a closure that returns a
    // dictionary.
    Constants("PI" to Math.PI)

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") { "Hello world! 👋" }

    AsyncFunction("generateCircomProof") { zkeyPath: String, circuitInputs: String, expoProofLib: ExpoCircomProofLib ->
      try {
        val file = File(zkeyPath)
    
        if (!file.exists()) {
          throw CodedException("ZkeyFileNotFound", "The zkey file was not found at path: $zkeyPath", null)
        }
    
        val proofLib = if (expoProofLib.proofLib == ProofLibOption.arkworks) ProofLib.ARKWORKS else ProofLib.RAPIDSNARK
        val res = generateCircomProof(file.absolutePath, circuitInputs, proofLib)
        ExpoCircomProofResult(convertCircomProof(res.proof), res.inputs)
      } catch (e: Exception) {
        throw CodedException("GenerateProofFailed", "Unknown error occurred during proof generation", e)
      }    
    }

    AsyncFunction("verifyCircomProof") { zkeyPath: String, proofResult: ExpoCircomProofResult, expoProofLib: ExpoCircomProofLib ->
      try {
        val file = File(zkeyPath)
        if (!file.exists()) {
          throw CodedException("ZkeyFileNotFound", "The zkey file was not found at path: $zkeyPath", null)
        }
        val proofLib = if (expoProofLib.proofLib == ProofLibOption.arkworks) ProofLib.ARKWORKS else ProofLib.RAPIDSNARK
        val isValid = verifyCircomProof(file.absolutePath, convertCircomProofResult(proofResult), proofLib)
        isValid
    } catch (e: Exception) {
      throw CodedException("VerifyProofFailed", "Unknown error occurred during proof verification", e)
    }   
    }

    AsyncFunction("generateHalo2Proof") { srsPath: String, pkPath: String, circuitInputs: Map<String, List<String>> ->
      try {
        val srsFile = File(srsPath)
        val pkFile = File(pkPath)
        if (!srsFile.exists()) {
          throw CodedException("SrsFileNotFound", "The srs file was not found at path: $srsPath", null)
        }
        if (!pkFile.exists()) {
          throw CodedException("PkFileNotFound", "The pk file was not found at path: $pkPath", null)
        }
        val res = generateHalo2Proof(srsFile.absolutePath, pkFile.absolutePath, circuitInputs)
        ExpoHalo2ProofResult(res.proof, res.inputs)
      } catch (e: Exception) {
        throw CodedException("GenerateProofFailed", "Unknown error occurred during proof generation", e)
      }   
    }

    AsyncFunction("verifyHalo2Proof") { srsPath: String, vkPath: String, proof: ByteArray, publicInput: ByteArray ->
      try {
        val srsFile = File(srsPath)
        val vkFile = File(vkPath)
        if (!srsFile.exists()) {
          throw CodedException("SrsFileNotFound", "The srs file was not found at path: $srsPath", null)
        }
        if (!vkFile.exists()) {
          throw CodedException("VkFileNotFound", "The vk file was not found at path: $vkPath", null)
        }
        val isValid = verifyHalo2Proof(srsFile.absolutePath, vkFile.absolutePath, proof, publicInput)
        isValid
        } catch (e: Exception) {
          throw CodedException("VerifyProofFailed", "Unknown error occurred during proof verification", e)
        }   
    }

    AsyncFunction("generateNoirProof") { circuitPath: String, srsPath: String?, inputs: List<String>, onChain: Boolean, vk: ByteArray, lowMemoryMode: Boolean ->
      try {
        val res = generateNoirProof(circuitPath, srsPath, inputs, onChain, vk, lowMemoryMode)
        res
      } catch (e: Exception) {
        throw CodedException("GenerateProofFailed", "Unknown error occurred during proof generation", e)
      }   
    }

    AsyncFunction("verifyNoirProof") { circuitPath: String, proof: ByteArray, onChain: Boolean, vk: ByteArray, lowMemoryMode: Boolean ->
      try {
        val res = verifyNoirProof(circuitPath, proof, onChain, vk, lowMemoryMode)
        res
      } catch (e: Exception) {
        throw CodedException("VerifyProofFailed", "Unknown error occurred during proof verification", e)
      }   
    }

    AsyncFunction("getNoirVerificationKey") { circuitPath: String, srsPath: String?, onChain: Boolean, lowMemoryMode: Boolean ->
      try {
        val res = getNoirVerificationKey(circuitPath, srsPath, onChain, lowMemoryMode)
        res
      } catch (e: Exception) {
        throw CodedException("GenerateProofFailed", "Unknown error occurred during verification key generation", e)
      }   
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("setValueAsync") { value: String ->
      // Send an event to JavaScript.
      sendEvent("onChange", mapOf("value" to value))
    }

    // Enables the module to be used as a native view. Definition components that are accepted as
    // part of
    // the view definition: Prop, Events.
    View(MoproView::class) {
      // Defines a setter for the `name` prop.
      Prop("name") { view: MoproView, prop: String -> println(prop) }
    }
  }
}
